scripts/nbconverted/process_xCell.py

```diff
@@ -0,0 +1,132 @@
+
+# coding: utf-8
+
+# # Process xCell cell-type genesets
+# 
+# The genesets were extracted from additional file 3 in [Aran et al. 2017](https://doi.org/10.1186/s13059-017-1349-1 "xCell: digitally portraying the tissue cellular heterogeneity landscape"). The data consist of 489 gene signatures. Many of these signatures represent redundant cell-types with different evidence.
+
+# In[1]:
+
+
+import os
+from datetime import datetime
+import csv
+import pandas as pd
+
+
+# In[2]:
+
+
+get_ipython().run_line_magic('matplotlib', 'inline')
+
+
+# In[3]:
+
+
+# Load Additional File 3
+xcell_file = os.path.join('data', '13059_2017_1349_MOESM3_ESM.xlsx')
+xcell_df = pd.read_excel(xcell_file, index_col=0)
+print(xcell_df.shape)
+xcell_df.head()
+
+
+# In[4]:
+
+
+# Distribution of number of cell types
+cell_types = pd.Series([x.split('_')[0] for x in xcell_df.index])
+cell_types.value_counts().head(10)
+
+
+# In[5]:
+
+
+# Distribution of number of cell types
+cell_types.value_counts().hist(bins=10)
+
+
+# In[6]:
+
+
+geneset_url = 'https://doi.org/10.1186/s13059-017-1349-1'
+
+
+# In[7]:
+
+
+# Load curated gene names from versioned resource 
+commit = '721204091a96e55de6dcad165d6d8265e67e2a48'
+url = 'https://raw.githubusercontent.com/cognoma/genes/{}/data/genes.tsv'.format(commit)
+gene_df = pd.read_table(url)
+
+# Only consider protein-coding genes
+gene_df = (
+    gene_df.query("gene_type == 'protein-coding'")
+    .drop_duplicates('entrez_gene_id')
+)
+
+symbol_to_entrez = dict(zip(gene_df.symbol,
+                            gene_df.entrez_gene_id))
+
+# Add alternative symbols to entrez mapping dictionary
```

Comment:
This is very tricky to do right. You need to make sure all approved symbols map and don't get masked by synonyms. Also you don't want to map a synonym if there's ambiguity (multiple genes have the same synonym).

Cognoma genes does this for there case where you have symbol and chromosome info to map from:

https://github.com/cognoma/genes/blob/ad9631bb4e77e2cdc5413b0d77cb8f7e93fc5bee/2.process.py#L109-L122

You could adapt that code for just the symbol case.

Reply:
Maybe dropping the row if a `symbol` was found in `synonyms` could also work

Reply:
In the synonym table, you `.drop_duplicates(['symbol'], keep=False)`. This means all ambigious instances are dropped. You then append this synonym table to the symbol table and drop_duplicates with keep=first

Reply:
the synonym table looks like:

entrez_gene_id | symbol | description | chromosome | gene_type | synonyms | aliases | all_synonyms
-- | -- | -- | -- | -- | -- | -- | --
1 | A1BG | alpha-1-B glycoprotein | 19 | protein-coding | [A1B, ABG, GAB, HYST2477] | alpha-1B-glycoprotein\|HEL-S-163pA\|epididymis s... | A1B
1 | A1BG | alpha-1-B glycoprotein | 19 | protein-coding | [A1B, ABG, GAB, HYST2477] | alpha-1B-glycoprotein\|HEL-S-163pA\|epididymis s... | ABG
1 | A1BG | alpha-1-B glycoprotein | 19 | protein-coding | [A1B, ABG, GAB, HYST2477] | alpha-1B-glycoprotein\|HEL-S-163pA\|epididymis s... | GAB
1 | A1BG | alpha-1-B glycoprotein | 19 | protein-coding | [A1B, ABG, GAB, HYST2477] | alpha-1B-glycoprotein\|HEL-S-163pA\|epididymis s... | HYST2477
2 | A2M | alpha-2-macroglobulin | 12 | protein-coding | [A2MD, CPAMD5, FWP007, S863-7] | alpha-2-macroglobulin\|C3 and PZP-like alpha-2-... | A2MD

so the solution of dropping `symbol` duplicates will remove most `synonym` to `entrez_gene_id` mappings. My goal was to search all symbols in a resource and map them to entrez. The trouble is, many of these symbols exist in the `synonym` column.

I understand the concern here though - often times a `synonym` will be "upgraded" to a `symbol` and thus have mutiple `entrez_gene_id` mappings. We want to avoid a `synonym` being mapped to an `entrez_gene_id` if there is a direct `symbol` to `entrez_gene_id`. Is this accurate?

Take, the gene _NAT1_ as an example:

entrez_gene_id | symbol | description | chromosome | gene_type | synonyms | aliases | all_synonyms
-- | -- | -- | -- | -- | -- | -- | --
9 | NAT1 | N-acetyltransferase 1 | 8 | protein-coding | [AAC1, MNAT, NAT-1, NATI] | arylamine N-acetyltransferase 1\|N-acetyltransf... | AAC1
9 | NAT1 | N-acetyltransferase 1 | 8 | protein-coding | [AAC1, MNAT, NAT-1, NATI] | arylamine N-acetyltransferase 1\|N-acetyltransf... | MNAT
9 | NAT1 | N-acetyltransferase 1 | 8 | protein-coding | [AAC1, MNAT, NAT-1, NATI] | arylamine N-acetyltransferase 1\|N-acetyltransf... | NAT-1
9 | NAT1 | N-acetyltransferase 1 | 8 | protein-coding | [AAC1, MNAT, NAT-1, NATI] | arylamine N-acetyltransferase 1\|N-acetyltransf... | NATI

But also:

```python
gene_with_syn_df.query('all_synonyms == "NAT1"')
```

entrez_gene_id | symbol | description | chromosome | gene_type | synonyms | aliases | all_synonyms
-- | -- | -- | -- | -- | -- | -- | --
1982 | EIF4G2 | eukaryotic translation initiation factor 4 gam... | 11 | protein-coding | [AAG1, DAP5, NAT1, P97] | eukaryotic translation initiation factor 4 gam... | NAT1
6530 | SLC6A2 | solute carrier family 6 member 2 | 16 | protein-coding | [NAT1, NET, NET1, SLC6A5] | sodium-dependent noradrenaline transporter\|neu... | NAT1
10991 | SLC38A3 | solute carrier family 38 member 3 | 3 | protein-coding | [G17, NAT1, SN1, SNAT3] | sodium-coupled neutral amino acid transporter ... | NAT1

A dictionary of `symbol` to `entrez_gene_id` should map _NAT1_ to "9".

Reply:
An issue is what to do with multiple `synonym` mappings...

```python
# Remove rows that have redundant symbols in all_synonyms
gene_with_syn_df = gene_with_syn_df.query('symbol not in all_synonyms')

gene_with_syn_df.shape
> (47446, 8)

gene_with_syn_df.drop_duplicates(['all_synonyms']).shape
> (44659, 8)
```

Reply:
Speaking with @dhimmel in person:

`.drop_duplicates(['synonym'], keep=False)` needs to be run first, then dropping rows where the `symbol` is found in the `synonyms` second